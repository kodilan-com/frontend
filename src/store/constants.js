export const TOGGLE_LOADING = 'TOGGLE_LOADING';
export const SET_RECENT_POSTS = 'SET_RECENT_POSTS';
export const SET_ALL_POSTS = 'SET_ALL_POSTS';
export const SET_FEATURED = 'SET_FEATURED';
export const SET_TAGS = 'SET_TAGS';
export const SET_AVAILABLE_LOCATIONS = 'SET_AVAILABLE_LOCATIONS';
export const SET_ACTIVE_PERIOD = 'SET_ACTIVE_PERIOD';
export const RECENT_POST_COUNT = 25;
export const JOB_TYPE_MAP = {
  0: 'Tam zamanlı',
  1: 'Tam zamanlı',
  2: 'Yarı zamanlı',
  3: 'Stajyer',
  4: 'Freelance',
};

export const PERIODS = [
  {
    type: 'daily',
    text: 'Bugün',
    slug: 'bugun',
  },
  {
    type: 'weekly',
    text: 'Bu hafta',
    slug: 'bu-hafta',
  },
  {
    type: 'monthly',
    text: 'Bu ay',
    slug: 'bu-ay',
    isDefault: true,
  },
];

export const JOB_TYPE_SLUG_TO_ID_MAP = {
  'tam-zamanli': 1,
  'yari-zamanli': 2,
  stajyer: 3,
  freelance: 4,
};

export const JOB_TYPE_ID_TO_SLUG_MAP = {
  1: 'tam-zamanli',
  2: 'yari-zamanli',
  3: 'stajyer',
  4: 'freelance',
};

export const JOB_TYPES_FOR_DROPDOWN = [
  {
    id: 1,
    text: 'Tam zamanlı',
  },
  {
    id: 2,
    text: 'Yarı zamanlı',
  },
  {
    id: 3,
    text: 'Stajyer',
  },
  {
    id: 4,
    text: 'Freelance',
  },
];
