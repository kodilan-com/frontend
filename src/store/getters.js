export default {
  autocompleteTags(state) {
    return state.tags.map(tag => ({ label: tag.name, value: tag.slug }));
  },
  companyId(state) {
    return Object.keys(state.me).length && state.me.companies[0].id;
  },
};
