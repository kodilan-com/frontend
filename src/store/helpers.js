export default {
  rankPosts(post, postTagSlugs, relatedPosts) {
    const DAY_IN_MS = 1000 * 60 * 60 * 24;
    const TODAY = new Date();
    const POINTS = {
      createdInTenDays: 2,
      fromSameCompany: 3,
      hasSamePostType: 4,
      hasSameTag: 5,
      inSameLocation: 6,
    };

    return relatedPosts
      .filter(p => p.slug !== post.slug)
      .map((p) => {
        let score = 0;

        score += p.location === post.location ? POINTS.inSameLocation : 0;
        score += p.company.slug === post.company.slug ? POINTS.fromSameCompany : 0;
        score += p.type === post.type ? POINTS.hasSamePostType : 0;
        score += (TODAY - new Date(p.updated_at)) / DAY_IN_MS <= 10 ? POINTS.createdInTenDays : 0;
        p.tags.forEach((t) => {
          score += postTagSlugs.indexOf(t.slug) > -1 ? POINTS.hasSameTag : 0;
        });

        return { score, post: p };
      })
      .sort((c, n) => (c.score > n.score ? -1 : 1))
      .slice(0, 3)
      .map(o => o.post);
  },
};
