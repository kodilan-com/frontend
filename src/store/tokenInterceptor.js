export default (acc) => {
  const token = localStorage.getItem('AccessToken');
  if (token !== null) {
    acc.headers.Authorization = `Bearer ${token}`;
  }
  return acc;
};
