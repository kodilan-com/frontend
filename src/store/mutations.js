import * as constants from './constants';

export default {
  [constants.TOGGLE_LOADING](state) {
    state.isLoading = !state.isLoading;
  },
  [constants.SET_RECENT_POSTS](state, recentPosts) {
    state.recentPosts = recentPosts;
  },
  [constants.SET_ALL_POSTS](state, allPosts) {
    state.allPosts = allPosts;
  },
  [constants.SET_ALL_POST_META](state, meta) {
    state.allPostsMeta = meta;
  },
  [constants.SET_FEATURED](state, featuredPosts) {
    state.featuredPosts = featuredPosts;
  },
  [constants.SET_TAGS](state, tags) {
    state.tags = tags;
  },
  [constants.SET_AVAILABLE_LOCATIONS](state, locations) {
    state.availableLocations = locations.reduce((arr, loc) => {
      if (loc.location !== 'Remote') {
        arr.push(loc.location);
      }

      return arr;
    }, ['Remote']);
  },
  [constants.SET_ACTIVE_PERIOD](state, period) {
    state.activePeriod = period;
  },
  setIsLoggedIn(state, data) {
    state.isLoggedIn = data;
  },
  SET_ME(state, data) {
    state.me = data;
  },
  postListing(state, data) {
    state.manageJobs = data;
  },
};
