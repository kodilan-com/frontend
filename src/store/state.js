import { PERIODS } from './constants';

export default {
  isLoggedIn: localStorage.getItem('AccessToken') || null,
  me: {},
  manageJobs: [],
  isLoading: true,
  recentPosts: [],
  activePeriod: PERIODS.find(period => period.isDefault).type,
  featuredPosts: [],
  allPosts: [],
  allPostsMeta: {},
  tags: [],
  availableLocations: ['Remote', 'İstanbul', 'İzmir', 'Ankara', 'Bursa'],
};
