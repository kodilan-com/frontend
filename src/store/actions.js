import axios from 'axios';
import { cacheAdapterEnhancer } from 'axios-extensions';
import * as constants from './constants';
import helpers from './helpers';

const http = axios.create({
  baseURL: 'https://api.kodilan.com',
  adapter: cacheAdapterEnhancer(axios.defaults.adapter),
});

export default {
  toggleLoading({ commit }) {
    commit(constants.TOGGLE_LOADING);
  },
  fetchRecentPosts({ state, commit }) {
    return http.get(`/posts?get=${constants.RECENT_POST_COUNT}&period=${state.activePeriod}`)
      .then((res) => {
        commit(constants.SET_RECENT_POSTS, res.data.data);

        return res.data;
      });
  },
  fetchFeaturedPosts({ commit }) {
    return http.get('/posts?get=3&is_featured=1')
      .then((res) => {
        commit(constants.SET_FEATURED, res.data.data);

        return res.data;
      });
  },
  fetchAllPosts({ commit }, payload = {}) {
    const page = payload
    && payload.page
    && Number.isInteger(payload.page)
    && payload.page > 1 ? payload.page : 1;

    return http.get(`/posts?get=${constants.PER_PAGE}&page=${page}`)
      .then((res) => {
        const { data } = res;
        commit(constants.SET_ALL_POSTS, data.data);
        commit(constants.SET_ALL_POST_META, {
          total: data.total,
          current_page: data.current_page,
          last_page: data.last_page,
        });

        return res.data;
      });
  },
  fetchBySlug(_, slug) {
    return http.get(`/posts/${slug}`)
      .then(res => res.data);
  },
  fetchByCompany(_, company) {
    return http.get(`/companies/${company}/posts`)
      .then(res => res.data);
  },
  fetchByTag(_, tag) {
    return http.get(`/tags/${tag}/posts`)
      .then(res => res.data);
  },
  search(_, params) {
    return http.get('/search', { params })
      .then(res => res.data);
  },
  fetchTags({ commit }) {
    return http.get('/tags')
      .then((res) => {
        commit(constants.SET_TAGS, res.data.data);
      });
  },
  fetchRelatedPosts({ dispatch }, post) {
    const postTags = post.tags.map(t => t.slug);
    const categoryTags = ['frontend', 'backend', 'mobile', 'designer', 'qa', 'devops'];
    const [mainTag] = postTags.filter(t => categoryTags.indexOf(t) > -1);

    if (!mainTag) {
      return Promise.resolve([]);
    }

    return dispatch('fetchByTag', mainTag)
      .then(res => helpers.rankPosts(post, postTags, res.data));
  },
  fetchAvailableLocations({ commit }) {
    return http.get('/posts/locations')
      .then((res) => {
        commit(constants.SET_AVAILABLE_LOCATIONS, res.data);
      });
  },
  savePost(_, data) {
    return http.post('/posts', data);
  },
  subscribe(_, data) {
    return http.post('/newsletters', data);
  },
  setPeriod({ commit }, period) {
    commit('SET_ACTIVE_PERIOD', period);
  },
};
