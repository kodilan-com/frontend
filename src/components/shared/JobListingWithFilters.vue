<script>
import JobListing from './JobListing';
import LocationSelect from './LocationSelect';
import queryUtils from '../../utils/query';
import { JOB_TYPES_FOR_DROPDOWN } from '../../store/constants';

export default {
  props: {
    posts: {
      type: Array,
      required: true,
    },
    meta: {
      type: Object,
      required: false,
    },
    isLoading: {
      type: Boolean,
      required: false,
      default: false,
    },
  },
  components: {
    JobListing,
    LocationSelect,
  },
  data() {
    return {
      params: {
        query: this.$route.query.query || '',
        location: this.$route.query.location || '',
        type: this.$route.query.type || 1,
      },
      type: null,
      typeOptions: JOB_TYPES_FOR_DROPDOWN,
    };
  },
  computed: {
    countText() {
      const length = this.meta && this.meta.total ? this.meta.total : this.posts.length;

      if (!length) {
        return 'Bu kritere uygun ilan bulunamadı.';
      }

      const prefix = this.$route.path === '/ilanlar' ? 'Toplam' : 'Bu kriterlere uygun';

      return `${prefix} ${length} ilan bulunuyor`;
    },
  },
  methods: {
    search() {
      this.params.type = this.type ? this.type.id : 0;

      const params = queryUtils.getParams(this.params);

      if (params) {
        this.$router.push({ path: '/ilan-ara', query: params });
      }
    },
  },
  created() {
    const type = parseInt(this.$route.query.type, 10);

    if (type) {
      this.type = {
        id: type,
        text: this.typeOptions.find(o => o.id === type).text,
      };
    }
  },
};
</script>

<template>
  <section id="page">
    <div v-if="!isLoading" id="titlebar">
      <div class="container">
        <div class="sixteen columns">
          <slot name="headerText" />
          <span>{{ countText }}</span>
        </div>
      </div>
    </div>
    <div class="container job-listing" :class="{ empty: !posts.length }">
      <div class="eleven columns">
        <div class="padding-right">
          <job-listing :is-loading="isLoading"
                       :posts="posts"
          />
          <slot name="paginate" />
        </div>
      </div>
      <div v-if="!isLoading" class="five columns">
        <div class="widget">
          <input
            v-model="params.query"
            @keypress.enter="search"
            type="text"
            class="ico-01"
            placeholder="Pozisyon adı, teknoloji adı"
          >
          <location-select v-model="params.location" :value="params.location" :searchable="true" />
          <multiselect
            v-model="type"
            :options="typeOptions"
            label="text"
            :searchable="false"
            :close-on-select="true"
            placeholder="Çalışma tipi seçiniz..."
          />
        </div>
        <div class="button-right">
          <button @click="search" class="button tag-search-btn">
            Ara
          </button>
        </div>
      </div>
    </div>
  </section>
</template>
