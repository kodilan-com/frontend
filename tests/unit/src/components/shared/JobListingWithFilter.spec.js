import { shallowMount, createLocalVue } from '@vue/test-utils';
import JobListingWithFilters from '@/components/shared/JobListingWithFilters';
import queryUtils from '@/utils/query';
import { JOB_TYPES_FOR_DROPDOWN } from '@/store/constants';

jest.mock('@/utils/query', () => ({
  getParams: jest.fn(),
}));

describe('JobListingWithFilters.vue', () => {
  let wrapper;
  const posts = [{
    slug: 'post1',
  }];
  const $router = {
    push: jest.fn(),
  };
  let $route;
  const localVue = createLocalVue();

  localVue.component('multiselect', {
    props: ['options', 'searchable', 'closeOnSelect'],
  });

  const renderComponent = (options) => {
    const componentOptions = {
      stubs: ['location-select', 'job-listing'],
      mocks: {
        $route,
        $router,
      },
      propsData: {
        posts,
      },
      localVue,
      ...options,
    };

    wrapper = shallowMount(JobListingWithFilters, componentOptions);
  };

  beforeEach(() => {
    $route = {
      query: {
        query: 'test_query',
        location: 'test_location',
        type: 'test_type',
      },
    };

    renderComponent();
  });

  afterEach(() => {
    wrapper.destroy();

    jest.restoreAllMocks();
    jest.clearAllMocks();
  });

  it('should have data which generated with route query params', () => {
    expect(wrapper.vm.$data).toEqual({
      params: {
        location: 'test_location',
        query: 'test_query',
        type: 'test_type',
      },
      type: null,
      typeOptions: JOB_TYPES_FOR_DROPDOWN,
    });
  });

  it('should have data which generated with fallbacks', () => {
    $route = {
      query: {},
    };

    renderComponent();

    expect(wrapper.vm.$data).toEqual({
      params: {
        location: '',
        query: '',
        type: 1,
      },
      type: null,
      typeOptions: JOB_TYPES_FOR_DROPDOWN,
    });
  });

  describe('titlebar', () => {
    it('should not render titlebar div', () => {
      renderComponent({
        propsData: {
          isLoading: true,
          posts,
        },
      });

      expect(wrapper.find('#titlebar').exists()).toBe(false);
    });

    it('should render titlebar div, if isLoading prop is false', () => {
      expect(wrapper.find('#titlebar').exists()).toBe(true);
    });

    it('should match snapshot with "Bu kritere uygun ilan bulunamadı." text if posts length is 0', () => {
      renderComponent({
        propsData: {
          isLoading: false,
          posts: [],
        },
      });

      expect(wrapper.find('#titlebar').element).toMatchSnapshot();
    });

    it('should match snapshot with "Toplam 1 ilan bulunuyor" text if posts length is not 0 and $route path is "/ilanlar"', () => {
      renderComponent({
        propsData: {
          isLoading: false,
          posts,
        },
        mocks: {
          $route: {
            ...$route,
            path: '/ilanlar',
          },
        },
      });

      expect(wrapper.find('#titlebar span').text()).toBe('Toplam 1 ilan bulunuyor');
    });

    it('should match snapshot with "Bu kriterlere uygun 1 ilan bulunuyor" text if posts length is not 0 and $route path is "/ilanlar"', () => {
      renderComponent({
        propsData: {
          isLoading: false,
          posts,
        },
      });

      expect(wrapper.find('#titlebar span').text()).toBe('Bu kriterlere uygun 1 ilan bulunuyor');
    });
  });

  it('should have type object if query type is a parsable integer', () => {
    $route.query.type = '1';

    renderComponent();

    expect(wrapper.vm.type).toEqual({
      id: 1,
      text: 'Tam zamanlı',
    });
  });

  it('should have null type object if query type is not a parsable integer', () => {
    $route.query.type = 'not_parsable';

    renderComponent();

    expect(wrapper.vm.type).toEqual(null);
  });

  it('should have "empty" class if posts length is zero', () => {
    renderComponent({
      propsData: {
        posts: [],
      },
    });

    expect(wrapper.find('.job-listing').is('.empty')).toBe(true);
  });

  it('should pass isLoading and posts to JobListing component', () => {
    renderComponent({
      propsData: {
        isLoading: true,
        posts,
      },
    });

    expect(wrapper.find('job-listing-stub').props('isLoading')).toBe(true);
    expect(wrapper.find('job-listing-stub').props('posts')).toEqual(posts);
  });

  describe('widget', () => {
    it('should not be rendered, if isLoading is true', () => {
      renderComponent({
        propsData: {
          isLoading: true,
          posts,
        },
      });

      expect(wrapper.find('.five.columns').exists()).toBe(false);
    });

    it('should be rendered, if isLoading is false', () => {
      expect(wrapper.find('.five.columns .widget').exists()).toBe(true);
    });

    it('should match snapshot input, if isLoading is false', () => {
      expect(wrapper.find('.five.columns input').element).toMatchSnapshot();
    });

    it('should change "query" data, if types anything', () => {
      const input = wrapper.find('.five.columns input');

      input.element.value = 'test';
      input.trigger('input');

      expect(wrapper.vm.params.query).toBe('test');
    });

    it('should search typed "query" data, if press enter', () => {
      const input = wrapper.find('.five.columns input');
      const searchQuery = 'test';

      queryUtils.getParams.mockReturnValue('testParams');

      input.element.value = searchQuery;
      input.trigger('input');
      input.trigger('keypress.enter');

      expect(queryUtils.getParams.mock.calls[0][0]).toEqual({
        type: 0,
        query: searchQuery,
        location: 'test_location',
      });
      expect($router.push).toHaveBeenCalled();
      expect($router.push.mock.calls[0][0]).toEqual({ path: '/ilan-ara', query: 'testParams' });
    });

    it('should not search, if getParams function returns falsy value', () => {
      const input = wrapper.find('.five.columns input');

      wrapper.setData({
        type: {
          id: 'test_id',
        },
      });

      queryUtils.getParams.mockReturnValue(undefined);

      input.trigger('keypress.enter');

      expect($router.push).not.toHaveBeenCalled();
      expect(queryUtils.getParams.mock.calls[0][0]).toEqual({
        type: 'test_id',
        query: $route.query.query,
        location: $route.query.location,
      });
    });

    it('should render LocationSelect component with correct props, if isLoading is false', () => {
      const locationSelect = wrapper.find('location-select-stub');

      expect(locationSelect.exists()).toBe(true);
      expect(locationSelect.props('searchable')).toBe(true);
      expect(locationSelect.props('value')).toBe('test_location');
    });

    it('should render MultiSelect component with correct props, if isLoading is false', () => {
      const multiSelect = wrapper.find('multiselect-stub');


      expect(multiSelect.element).toMatchSnapshot();
      expect(multiSelect.props('closeOnSelect')).toBe(true);
      expect(multiSelect.props('searchable')).toBe(false);
      expect(multiSelect.props('options')).toEqual(JOB_TYPES_FOR_DROPDOWN);
    });

    it('should render search button, if isLoading is false', () => {
      const searchButton = wrapper.find('.button-right button');

      expect(searchButton.element).toMatchSnapshot();
    });

    it('should search, if search button is clicked when isLoading is false', () => {
      const searchButton = wrapper.find('.button-right button');

      queryUtils.getParams.mockReturnValue('testParams');

      searchButton.trigger('click');

      expect($router.push).toHaveBeenCalled();
      expect($router.push.mock.calls[0][0]).toEqual({ path: '/ilan-ara', query: 'testParams' });
    });
  });
});
