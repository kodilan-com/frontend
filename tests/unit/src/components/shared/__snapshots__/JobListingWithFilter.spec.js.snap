// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobListingWithFilters.vue titlebar should match snapshot with "Bu kritere uygun ilan bulunamadı." text if posts length is 0 1`] = `
<div
  id="titlebar"
>
  <div
    class="container"
  >
    <div
      class="sixteen columns"
    >
       
      <span>
        Bu kritere uygun ilan bulunamadı.
      </span>
    </div>
  </div>
</div>
`;

exports[`JobListingWithFilters.vue widget should match snapshot input, if isLoading is false 1`] = `
<input
  class="ico-01"
  placeholder="Pozisyon adı, teknoloji adı"
  type="text"
/>
`;

exports[`JobListingWithFilters.vue widget should render MultiSelect component with correct props, if isLoading is false 1`] = `
<multiselect-stub
  closeonselect="true"
  label="text"
  options="[object Object],[object Object],[object Object],[object Object]"
  placeholder="Çalışma tipi seçiniz..."
/>
`;

exports[`JobListingWithFilters.vue widget should render search button, if isLoading is false 1`] = `
<button
  class="button tag-search-btn"
>
  
          Ara
        
</button>
`;
